### PR TITLE
Cryptocurrency demo

### PR DESCRIPTION
**Pull request by KateNegrienko**
## Overview

Create new cryprocurrency demo for java binding

---
See: https://jira.bf.local/browse/ECR-1773

### CODE REVIEW

File: exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        res.json(body);
+      });
+    
+    
+  }
+
+});
+
+module.exports = router;
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js]: EOL

File: exonum-java-binding-cryptocurrency-demo/frontend/src/pages/Auth.vue [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    methods: {
+      login() {
+        if (!this.$validateHex(this.secretKey, 64)) {
+          return this.$notify('error', 'Invalid secret key is passed')
+        }
+
+        this.isSpinnerVisible = true
+
+        this.$store.commit('login', {
+          publicKey: this.secretKey.substr(64),
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/pages/Auth.vue]: Not sure I understand what happens here: do we store the public key in `this.secretKey` starting at position 64?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+let TransferTransactionProtobuf = new Type("TransferTransaction").add(new Field("seed", 1, "u-int64"));
+TransferTransactionProtobuf.add(new Field("senderId", 2, "bytes"))
+TransferTransactionProtobuf.add(new Field("recipientId", 3, "bytes"))
+TransferTransactionProtobuf.add(new Field("amount", 4, "int64"))
+
+var root = new Root();
+root.define("CreateTransactionProtobuf").add(CreateTransactionProtobuf);
+root.define("TransferTransactionProtobuf").add(TransferTransactionProtobuf);
+
+const TX_URL = '/api/cryptocurrency-demo-service/submit-transaction'// '/api/services/cryptocurrency/v1/wallets/transaction'
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: Is the commented code needed?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+module.exports = {
+  install(Vue) {
+    Vue.prototype.$blockchain = {
+      generateKeyPair() {
+        return Exonum.keyPair()
+      },
+
+      generateSeed() {
+        const buffer = nacl.randomBytes(7)
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: What is 7 — a seed, size, anything else? I'd suggest giving this constant a name.

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        return bigInt.fromArray(Array.from(buffer), 256).toString()
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: 256?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+         service_id: SERVICE_ID,
+         message_id: TX_WALLET_ID,
+         signature: signature,
+         body: data
+        })
+      },
+
+      transfer(keyPair, receiver, amountToTransfer, seed) {
+        
+        let buffer = MessageHead.serialize({
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: Wondering if it makes sense to extract a method `createMessageHeader(transactionId)` to remove duplicated code :thinking:

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        data.senderId = Exonum.uint8ArrayToHexadecimal(data.senderId)
+        data.recipientId = Exonum.uint8ArrayToHexadecimal(data.recipientId)
+
+        return axios.post(TX_URL, {
+          protocol_version: PROTOCOL_VERSION,
+          service_id: SERVICE_ID,
+          message_id: TX_TRANSFER_ID,
+          signature: signature,
+          body: data
+        }).then(response => waitForAcceptance(keyPair.publicKey, response.data))
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: Is it intentional that here we `waitForAcceptance`, but do not in the method above?

File: exonum-java-binding-cryptocurrency-demo/frontend/tests/blockchain.test.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      secretKey: 'e6361a3221d68148bcb5875af57b2ef8f3cbf821dadaab8ad27bc50d29abf22a9f3ed8007937950d889981a1f0beff041f54d704840e01c207b07b0d5581db13'
+    }
+    const receiver = '7c3cd965f8084b5730f0f95da1f3b5baf7554c044078d986e249d78c4ee00a98'
+    const amountToTransfer = '25'
+    const seed = '21093588264774074'
+    const data = await Vue.prototype.$blockchain.transfer(keyPair, receiver, amountToTransfer, seed)
+
+    expect(data).toEqual({"type": "committed"})
+  })
+})
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/tests/blockchain.test.js]: eol

File: exonum-java-binding-cryptocurrency-demo/frontend/src/pages/Auth.vue [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+          secretKey: this.secretKey
+        })
+
+        this.$nextTick(function() {
+          this.$router.push({ name: 'user' })
+        })
+      },
+
+      async register() {
+
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/pages/Auth.vue]: empty line

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 10]
```diff
@@ -0,0 +1,184 @@
+import * as Exonum from 'exonum-client'
+import axios from 'axios'
+import * as Protobuf from 'protobufjs/light'
+import bigInt from 'big-integer'
+import nacl from 'tweetnacl'
+
+var Root  = Protobuf.Root,
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: why does it `var` not `let`?

File: exonum-java-binding-cryptocurrency-demo/frontend/tests/blockchain.test.js [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+import actual from './data/actual.json'
+const mock = new MockAdapter(axios)
+const bigIntRegex = /[0-9]+/i;
+const hexRegex = /[0-9A-Fa-f]+/i;
+
+Vue.use(Blockchain)
+
+mock.onGet('/api/services/configuration/v1/configs/actual').reply(200, actual)
+
+
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/frontend/tests/blockchain.test.js]: redundant line?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/pages/Auth.vue (hunk 2 above)
Comment (boguslavsky) [reply exonum-java-binding-cryptocurrency-demo/frontend/src/pages/Auth.vue]: Yes, you are right. Here is example of signing key pair:
```
const keyPair = {
  publicKey: 'fa7f9ee43aff70c879f80fa7fd15955c18b98c72310b09e7818310325050cf7a',
  secretKey: '978e3321bd6331d56e5f4c2bdb95bf471e95a77a6839e68d4241e7b0932ebe2b' +
  'fa7f9ee43aff70c879f80fa7fd15955c18b98c72310b09e7818310325050cf7a'
}
```

File: exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
+const express = require('express');
+const request = require('request');
+const router = express.Router();
+
+router.get('/*', function (req, res, next) {
+  const query = req.params[0];
+
+console.log(query, query.indexOf("explorer"))
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js]: indents

File: exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+      }
+      try {
+        res.json(JSON.parse(body));
+      } catch (e) {
+        res.json({});
+      }
+    });
+    
+  }
+
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js]: empty lines

File: exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+});
+
+router.post('/*', function (req, res, next) {
+  const query = req.params[0];
+
+console.log(query , query.indexOf("cryptocurrency-demo-service"))
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js]: indents

File: exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        url: req.app.get('explorerRoot') + '/api/' + query,
+        json: req.body
+      },
+      function (err, response, body) {
+        if (err) {
+          return next(err);
+        }
+        res.json(body);
+      });
+
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/frontend/routes/api.js]: empty lines

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js (hunk 10 above)
Comment (KateNegrienko) [reply exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: @bullet-tooth  I change to const :)

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js (hunk 5 above)
Comment (KateNegrienko) [reply exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: @dmitry-timofeev bytes

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js (hunk 7 above)
Comment (boguslavsky) [reply exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: @dmitry-timofeev yes, expected behavior

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 16]
```diff
@@ -111,13 +110,16 @@ module.exports = {
 
         const body = CreateTransactionProtobuf.encode(data).finish();
 
+        // calculate payload and insert it into buffer
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: Payload size or payload length or, better, message size?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 17]
```diff
@@ -111,13 +110,16 @@ module.exports = {
...
+        Exonum.Uint32.serialize(body.length + SIGNATURE_LENGTH, buffer, 6)
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: Don't we include the size of the header in the 'payload length': https://exonum.com/doc/architecture/serialization/#payload-length?

Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: 6 — `PAYLOAD_LENGTH_OFFSET` (if that's what it is)?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js (hunk 4 above)
Comment (boguslavsky) [reply exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: @dmitry-timofeev honestly it should be random integer of `8` bytes here, which is 2^64, but as soon as Java supports only 2^63 integers we decided to do this hack..

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+
+const TX_URL = '/api/cryptocurrency-demo-service/submit-transaction'
+const ATTEMPTS = 10
+const ATTEMPT_TIMEOUT = 500
+const PROTOCOL_VERSION = 0
+const SERVICE_ID = 42
+const TX_TRANSFER_ID = 2
+const TX_WALLET_ID = 1
+const SIGNATURE_LENGTH = 64
+const PAYLOD_SIZE_OFFSET = 6
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: `PAYLOAD_SIZE_OFFSET`

File: exonum-java-binding-cryptocurrency-demo/frontend/src/App.vue [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    <router-view/>
+
+    <footer class="pb-4 hr">
+      <hr class="mt-5 mb-5">
+      <div class="container">
+        <div class="row">
+          <div class="col-sm-12">
+            <img src="images/exonum.png" width="41" height="36" class="float-left mt-sm-1 mr-3" alt="">
+            <ul class="list-unstyled">
+              <li>Sources on <a href="https://github.com/exonum/cryptocurrency-advanced" target="_blank">GitHub</a></li>
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/App.vue]: `https://github.com/exonum/exonum-java-binding`

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+    { name: 'message_id', type: Exonum.Uint16 },
+    { name: 'service_id', type: Exonum.Uint16 },
+    { name: 'payload', type: Exonum.Uint32 }
+  ]
+})
+
+function getWallet(publicKey) {
+  return axios.get('/api/services/configuration/v1/configs/actual').then(response => {
+    // actual list of public keys of validators
+    const validators = response.data.config.validator_keys.map(validator => {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: Why do we make the request above? It seems `validators` are not used here.

**Comment by dmitry-timofeev** [comment]
Thanks, merged, further improvements in separate PRs.

**Pull request merged**